File: py/elo/transport/tcp.py
"""TCP connection manager — persistent peer connections with framing."""

from __future__ import annotations

import asyncio
import logging
import time
from collections.abc import Callable, Awaitable
from typing import Any

from elo.transport.protocol import (
    encode_frame,
    read_frame,
    write_frame,
    bye_msg,
    heartbeat_msg,
    FrameError,
    MessageType,
)

logger = logging.getLogger("elo.transport.tcp")

# Message handler signature: (peer_addr: str, message: dict) → None
MessageHandler = Callable[[str, dict[str, Any]], Awaitable[None]]


class PeerConnection:
    """A single TCP connection to a peer."""

    def __init__(self, addr: str, reader: asyncio.StreamReader,
                 writer: asyncio.StreamWriter):
        self.addr = addr
        self._reader = reader
        self._writer = writer
        self._alive = True
        self._last_read = time.monotonic()

    async def send(self, message: dict[str, Any]) -> None:
        """Send a framed message to the peer."""
        if not self._alive:
            raise ConnectionError("Peer disconnected")
        try:
            await write_frame(self._writer, message)
        except Exception:
            self._alive = False
            raise

    async def recv(self) -> dict[str, Any] | None:
        """Receive one framed message from the peer. Returns None on EOF."""
        try:
            msg = await read_frame(self._reader)
            self._last_read = time.monotonic()
            return msg
        except (FrameError, asyncio.IncompleteReadError, ConnectionError):
            self._alive = False
            return None

    def close(self) -> None:
        """Close the connection."""
        self._alive = False
        try:
            self._writer.close()
        except Exception:
            pass

    @property
    def alive(self) -> bool:
        return self._alive

    @property
    def idle_seconds(self) -> float:
        return time.monotonic() - self._last_read


class TCPManager:
    """Manages all TCP peer connections: server + client connections."""

    # Timeout configuration
    CONNECT_TIMEOUT = 10  # seconds
    HEARTBEAT_INTERVAL = 30  # seconds
    HEARTBEAT_IDLE_TIMEOUT = 90  # 3 missed heartbeats

    def __init__(self, node_id: str, host: str = "0.0.0.0", port: int = 7878):
        self._node_id = node_id
        self._host = host
        self._port = port
        self._server: asyncio.Server | None = None
        self._peers: dict[str, PeerConnection] = {}
        self._handler: MessageHandler | None = None
        self._shutdown = asyncio.Event()
        self._heartbeat_task: asyncio.Task | None = None
        self._listen_task: asyncio.Task | None = None

    # ── lifecycle ────────────────────────────────────────────

    async def start(self) -> int:
        """Start the TCP server. Returns the actual port."""
        self._server = await asyncio.start_server(
            self._handle_incoming,
            host=self._host,
            port=self._port,
        )
        # Get the actual port (useful if port=0 for random assignment)
        if self._port == 0 and self._server.sockets:
            self._port = self._server.sockets[0].getsockname()[1]

        logger.info("[tcp] listening on %s:%d", self._host, self._port)
        return self._port

    async def stop(self) -> None:
        """Stop the server and close all peer connections."""
        self._shutdown.set()

        if self._heartbeat_task:
            self._heartbeat_task.cancel()

        # Send BYE to all peers
        for addr, peer in list(self._peers.items()):
            try:
                await peer.send(bye_msg())
            except Exception:
                pass
            peer.close()

        self._peers.clear()

        if self._server:
            self._server.close()
            await self._server.wait_closed()

        logger.info("[tcp] stopped")

    # ── handlers ─────────────────────────────────────────────

    def on_message(self, handler: MessageHandler) -> None:
        """Register a message handler."""
        self._handler = handler

    async def _handle_incoming(self, reader: asyncio.StreamReader,
                               writer: asyncio.StreamWriter) -> None:
        """Handle a new incoming TCP connection."""
        peername = writer.get_extra_info("peername")
        addr = f"{peername[0]}:{peername[1]}" if peername else "unknown"
        logger.debug("[tcp] incoming from %s", addr)

        peer = PeerConnection(addr, reader, writer)

        try:
            # Read HELLO as first message
            hello = await peer.recv()
            if hello is None or hello.get("type") != MessageType.HELLO:
                logger.warning("[tcp] no hello from %s", addr)
                peer.close()
                return

            remote_id = hello.get("node_id", "")
            addr = f"{remote_id[:12]}@{addr}"
            peer.addr = addr

            self._peers[addr] = peer
            logger.info("[tcp] peer connected: %s", addr)

            # Notify handler about the peer + its HELLO
            if self._handler:
                await self._handler(addr, hello)

        except Exception:
            logger.debug("[tcp] connection error from %s", addr, exc_info=True)
            peer.close()
            return

        # Start read loop (shared by inbound and outbound)
        await self._read_loop(addr, peer)

    # ── read loop (shared by inbound + outbound) ───────────

    async def _read_loop(self, addr: str, peer: PeerConnection) -> None:
        """Continuously read messages from a peer until disconnect."""
        try:
            while not self._shutdown.is_set():
                msg = await peer.recv()
                if msg is None:
                    break
                if self._handler:
                    await self._handler(addr, msg)
        except Exception:
            logger.debug("[tcp] read error from %s", addr, exc_info=True)
        finally:
            peer.close()
            self._peers.pop(addr, None)
            if self._handler:
                try:
                    await self._handler(addr, {"type": MessageType.BYE, "node_id": ""})
                except Exception:
                    pass
            logger.info("[tcp] peer disconnected: %s", addr)

    # ── outbound connections ─────────────────────────────────

    async def connect_to_peer(self, addr: str, node_id: str = "unknown",
                               hello_payload: dict | None = None) -> str | None:
        """Connect to a remote peer. Returns the assigned peer address on success."""
        if ":" not in addr:
            logger.warning("[tcp] invalid address: %s", addr)
            return None

        # Normalize: if no port, use default
        host, _, port_str = addr.partition(":")
        port = int(port_str) if port_str else self._port
        canonical = f"{host}:{port}"

        if canonical in self._peers:
            return canonical

        try:
            reader, writer = await asyncio.wait_for(
                asyncio.open_connection(host, port),
                timeout=self.CONNECT_TIMEOUT,
            )
        except Exception as e:
            logger.debug("[tcp] connect to %s failed: %s", canonical, e)
            return None

        peer = PeerConnection(canonical, reader, writer)

        # Send HELLO
        if hello_payload:
            await peer.send(hello_payload)

        # Wait for HELLO_ACK
        try:
            ack = await asyncio.wait_for(peer.recv(), timeout=5)
        except asyncio.TimeoutError:
            peer.close()
            return None

        if ack is None or ack.get("type") != MessageType.HELLO_ACK:
            peer.close()
            return None

        remote_id = ack.get("node_id", "")
        display_addr = f"{remote_id[:12]}@{canonical}"
        peer.addr = display_addr

        self._peers[display_addr] = peer

        # Notify handler with HELLO_ACK
        if self._handler:
            await self._handler(display_addr, ack)

        logger.info("[tcp] connected to %s", display_addr)

        # Start read loop in background
        asyncio.create_task(self._read_loop(display_addr, peer))

        return display_addr

    async def send_to(self, peer_addr: str, message: dict[str, Any]) -> None:
        """Send a message to a specific peer."""
        peer = self._peers.get(peer_addr)
        if peer is None:
            raise ConnectionError(f"Peer not connected: {peer_addr}")
        await peer.send(message)

    # ── broadcast ────────────────────────────────────────────

    async def broadcast(self, message: dict[str, Any],
                        exclude: set[str] | None = None) -> None:
        """Send a message to all connected peers."""
        exclude = exclude or set()
        for addr, peer in list(self._peers.items()):
            if addr in exclude:
                continue
            try:
                await peer.send(message)
            except Exception:
                logger.debug("[tcp] broadcast failed to %s", addr)

    # ── heartbeat ────────────────────────────────────────────

    async def start_heartbeat(self) -> None:
        """Periodically send heartbeats and purge dead peers."""
        while not self._shutdown.is_set():
            try:
                await asyncio.wait_for(
                    self._shutdown.wait(),
                    timeout=self.HEARTBEAT_INTERVAL,
                )
                break
            except asyncio.TimeoutError:
                pass

            hb = heartbeat_msg(self._node_id)
            dead = []

            for addr, peer in list(self._peers.items()):
                try:
                    await peer.send(hb)
                except Exception:
                    dead.append(addr)

                # Check idle timeout
                if peer.idle_seconds > self.HEARTBEAT_IDLE_TIMEOUT:
                    dead.append(addr)

            for addr in set(dead):
                peer = self._peers.pop(addr, None)
                if peer:
                    peer.close()
                    if self._handler:
                        await self._handler(addr,
                                          {"type": MessageType.BYE, "node_id": ""})
                    logger.info("[tcp] peer timed out: %s", addr)

    # ── accessors ────────────────────────────────────────────

    @property
    def peer_count(self) -> int:
        return len(self._peers)

    @property
    def peer_addresses(self) -> list[str]:
        return list(self._peers.keys())

    @property
    def port(self) -> int:
        return self._port

File: go/transport/tcp.go
package transport

import (
	"fmt"
	"io"
	"log"
	"net"
	"sync"
	"time"
)

// PeerConn represents a single TCP connection to a peer.
type PeerConn struct {
	Addr     string
	conn     net.Conn
	mu       sync.Mutex
	lastRead time.Time
	alive    bool
}

// Send writes a framed message to the peer.
func (p *PeerConn) Send(msg map[string]any) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if !p.alive {
		return fmt.Errorf("peer disconnected")
	}
	frame, err := EncodeFrame(msg)
	if err != nil {
		return err
	}
	_, err = p.conn.Write(frame)
	return err
}

// Recv reads a framed message from the peer.
func (p *PeerConn) Recv() (map[string]any, error) {
	msg, err := ReadFrame(p.conn)
	if err != nil {
		p.alive = false
		return nil, err
	}
	p.lastRead = time.Now()
	return msg, nil
}

// Close closes the connection.
func (p *PeerConn) Close() {
	p.alive = false
	p.conn.Close()
}

// Idle returns the time since the last read.
func (p *PeerConn) Idle() time.Duration {
	return time.Since(p.lastRead)
}

// MessageHandler is the callback for incoming messages: (peerAddr, message).
type MessageHandler func(peerAddr string, msg map[string]any)

// TCPManager manages all TCP peer connections.
type TCPManager struct {
	nodeID   string
	host     string
	port     int
	listener net.Listener
	peers    map[string]*PeerConn
	mu       sync.RWMutex
	handler  MessageHandler
	shutdown chan struct{}
}

// NewTCPManager creates a new TCP connection manager.
func NewTCPManager(nodeID string, host string, port int) *TCPManager {
	return &TCPManager{
		nodeID:   nodeID,
		host:     host,
		port:     port,
		peers:    make(map[string]*PeerConn),
		shutdown: make(chan struct{}),
	}
}

// Start starts the TCP server.
func (m *TCPManager) Start() (int, error) {
	addr := fmt.Sprintf("%s:%d", m.host, m.port)
	l, err := net.Listen("tcp", addr)
	if err != nil {
		return 0, fmt.Errorf("listen: %w", err)
	}
	m.listener = l
	m.port = l.Addr().(*net.TCPAddr).Port

	go m.acceptLoop()
	log.Printf("[tcp] listening on %s:%d", m.host, m.port)
	return m.port, nil
}

// Stop shuts down the server and closes all peer connections.
func (m *TCPManager) Stop() {
	close(m.shutdown)
	if m.listener != nil {
		m.listener.Close()
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	for _, p := range m.peers {
		p.Close()
	}
	m.peers = make(map[string]*PeerConn)
}

// OnMessage registers the message handler.
func (m *TCPManager) OnMessage(handler MessageHandler) {
	m.handler = handler
}

// Port returns the listening port.
func (m *TCPManager) Port() int { return m.port }

// PeerCount returns the number of connected peers.
func (m *TCPManager) PeerCount() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.peers)
}

// PeerAddresses returns all connected peer addresses.
func (m *TCPManager) PeerAddresses() []string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	addrs := make([]string, 0, len(m.peers))
	for a := range m.peers {
		addrs = append(addrs, a)
	}
	return addrs
}

// HasPeer checks if a peer is connected.
func (m *TCPManager) HasPeer(addr string) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	_, ok := m.peers[addr]
	return ok
}

// SendTo sends a message to a specific peer.
func (m *TCPManager) SendTo(peerAddr string, msg map[string]any) error {
	m.mu.RLock()
	p, ok := m.peers[peerAddr]
	m.mu.RUnlock()
	if !ok {
		return fmt.Errorf("peer not found: %s", peerAddr)
	}
	return p.Send(msg)
}

// Broadcast sends a message to all peers.
func (m *TCPManager) Broadcast(msg map[string]any, exclude map[string]bool) error {
	m.mu.RLock()
	defer m.mu.RUnlock()
	for addr, p := range m.peers {
		if exclude != nil && exclude[addr] {
			continue
		}
		if err := p.Send(msg); err != nil {
			log.Printf("[tcp] broadcast failed to %s: %v", addr, err)
		}
	}
	return nil
}

// ConnectToPeer connects to a remote peer.
func (m *TCPManager) ConnectToPeer(addr string, hello map[string]any) (string, error) {
	m.mu.RLock()
	if _, exists := m.peers[addr]; exists {
		m.mu.RUnlock()
		return addr, nil
	}
	m.mu.RUnlock()

	conn, err := net.DialTimeout("tcp", addr, 10*time.Second)
	if err != nil {
		return "", fmt.Errorf("dial %s: %w", addr, err)
	}

	peer := &PeerConn{
		Addr:     addr,
		conn:     conn,
		alive:    true,
		lastRead: time.Now(),
	}

	// Send HELLO
	if hello != nil {
		if err := peer.Send(hello); err != nil {
			conn.Close()
			return "", err
		}
	}

	// Read HELLO_ACK
	conn.SetReadDeadline(time.Now().Add(5 * time.Second))
	ack, err := peer.Recv()
	conn.SetReadDeadline(time.Time{})
	if err != nil || ack["type"] != MsgHelloAck {
		conn.Close()
		return "", fmt.Errorf("no hello_ack from %s", addr)
	}

	m.mu.Lock()
	m.peers[addr] = peer
	m.mu.Unlock()

	// Start read loop
	go m.readLoop(addr, peer)

	return addr, nil
}

func (m *TCPManager) acceptLoop() {
	for {
		select {
		case <-m.shutdown:
			return
		default:
		}
		conn, err := m.listener.Accept()
		if err != nil {
			select {
			case <-m.shutdown:
				return
			default:
				continue
			}
		}
		go m.handleConn(conn)
	}
}

func (m *TCPManager) handleConn(conn net.Conn) {
	addr := conn.RemoteAddr().String()
	peer := &PeerConn{
		Addr:     addr,
		conn:     conn,
		alive:    true,
		lastRead: time.Now(),
	}

	// Expect HELLO first
	conn.SetReadDeadline(time.Now().Add(10 * time.Second))
	hello, err := peer.Recv()
	conn.SetReadDeadline(time.Time{})
	if err != nil || hello["type"] != MsgHello {
		conn.Close()
		return
	}

	m.mu.Lock()
	m.peers[addr] = peer
	m.mu.Unlock()

	if m.handler != nil {
		m.handler(addr, hello)
	}

	m.readLoop(addr, peer)
}

func (m *TCPManager) readLoop(addr string, peer *PeerConn) {
	for {
		peer.conn.SetReadDeadline(time.Now().Add(90 * time.Second))
		msg, err := peer.Recv()
		if err != nil {
			if err == io.EOF || !peer.alive {
				break
			}
			continue
		}
		if m.handler != nil {
			m.handler(addr, msg)
		}
	}

	m.mu.Lock()
	delete(m.peers, addr)
	m.mu.Unlock()
	peer.Close()
	if m.handler != nil {
		m.handler(addr, ByeMsg())
	}
}
